prov/efa/src/efa_msg.c

```diff
@@ -0,0 +1,408 @@
+/* SPDX-License-Identifier: BSD-2-Clause OR GPL-2.0-only */
+/* SPDX-FileCopyrightText: Copyright (c) 2013-2015 Intel Corporation, Inc.  All rights reserved. */
+/* SPDX-FileCopyrightText: Copyright Amazon.com, Inc. or its affiliates. All rights reserved. */
+
+#include "config.h"
+
+
+#include "ofi.h"
+#include "ofi_enosys.h"
+#include "ofi_iov.h"
+
+#include "efa_dgram_ep.h"
+#include "efa.h"
+#include "efa_av.h"
+
+#include "efa_tp.h"
+
+#define EFA_SETUP_IOV(iov, buf, len)           \
+	do {                                   \
+		iov.iov_base = (void *)buf;    \
+		iov.iov_len = (size_t)len;     \
+	} while (0)
+
+#define EFA_SETUP_MSG(msg, iov, _desc, count, _addr, _context, _data)    \
+	do {                                                             \
+		msg.msg_iov = (const struct iovec *)iov;                 \
+		msg.desc = (void **)_desc;                               \
```

Comment:
So this is not a copy of the user request (iov, desc), then and then if we use `ofi_consume_iov` , we will modify user's iov and desc, which is wrong.

So we either need to do a memcpy here, or keep using `ofi_iov_locate`